src/individual/BattleController_BeforeMove.c

```diff
@@ -0,0 +1,4354 @@
+            LoadBattleSubSeqScript(ctx, ARC_BATTLE_SUB_SEQ, SUB_SEQ_RESTORE_HP_FULL_FAIL);
+            ctx->next_server_seq_no = CONTROLLER_COMMAND_25;
+            ctx->server_seq_no = CONTROLLER_COMMAND_RUN_SCRIPT;
+            ctx->waza_status_flag |= MOVE_STATUS_NO_MORE_WORK;
+            ctx->wb_seq_no = BEFORE_MOVE_START;
+            return TRUE;
+        }
+        // Rest while user has Insomnia / Vital Spirit
+        if (attackerAbility == ABILITY_INSOMNIA || attackerAbility == ABILITY_VITAL_SPIRIT) {
+            ctx->oneTurnFlag[ctx->attack_client].parental_bond_flag = 0;
+            ctx->oneTurnFlag[ctx->attack_client].parental_bond_is_active = FALSE;
+            // TODO: test
+            ctx->battlerIdTemp = ctx->attack_client;
+            LoadBattleSubSeqScript(ctx, ARC_BATTLE_SUB_SEQ, SUB_SEQ_STAYED_AWAKE);
+            ctx->next_server_seq_no = CONTROLLER_COMMAND_25;
+            ctx->server_seq_no = CONTROLLER_COMMAND_RUN_SCRIPT;
+            ctx->waza_status_flag |= MOVE_STATUS_NO_MORE_WORK;
+            ctx->wb_seq_no = BEFORE_MOVE_START;
+            return TRUE;
+        }
+    }
+    return FALSE;
+}
+
+void BattleController_CheckBide(struct BattleSystem *bsys, struct BattleStruct *ctx) {
+    if (ctx->battlemon[ctx->attack_client].condition2 & STATUS2_BIDE) {
+        ctx->battlemon[ctx->attack_client].condition2 -= (1 << STATUS2_BIDE_SHIFT);
+        if (!(ctx->battlemon[ctx->attack_client].condition2 & STATUS2_BIDE) && ctx->store_damage[ctx->attack_client]) {
+            ctx->damage = ctx->store_damage[ctx->attack_client] * 2;
+            if (ctx->battlemon[ctx->client_no_hit[ctx->attack_client]].hp != 0) {
+                ctx->defence_client = ctx->client_no_hit[ctx->attack_client];
+            } else {
+                ctx->defence_client = Battler_GetRandomOpposingBattlerId(bsys, ctx, ctx->attack_client);
+                if (ctx->battlemon[ctx->defence_client].hp == 0) {
+                    LoadBattleSubSeqScript(ctx, ARC_BATTLE_SUB_SEQ, SUB_SEQ_BIDE_END_NO_TARGET);
+                    ctx->next_server_seq_no = CONTROLLER_COMMAND_39;
+                    ctx->server_seq_no = CONTROLLER_COMMAND_RUN_SCRIPT;
+                }
+            }
+        }
+        LoadBattleSubSeqScript(ctx, ARC_BATTLE_SUB_SEQ, SUB_SEQ_END_BIDE);
+        ctx->next_server_seq_no = ctx->server_seq_no;
+        ctx->server_seq_no = CONTROLLER_COMMAND_RUN_SCRIPT;
+    }
+}
+
+BOOL BattleController_CheckAbilityFailures1(struct BattleSystem *bsys UNUSED, struct BattleStruct *ctx) {
+    int attacker = ctx->attack_client;
+    int defender = ctx->defence_client;
+
+    // Handle Damp
+    switch (ctx->current_move_index) {
+    case MOVE_SELF_DESTRUCT:
+    case MOVE_EXPLOSION:
+    case MOVE_MIND_BLOWN:
+    case MOVE_MISTY_EXPLOSION:
+        if (CheckSideAbility(gBattleSystem, ctx, CHECK_ABILITY_ALL_HP, defender, ABILITY_DAMP)
+        && CLIENT_DOES_NOT_HAVE_MOLD_BREAKER_VARIATIONS(ctx, attacker)) {
```

Comment:
this is just something like
```c
if (MoldBreakerAbilityCheck(ctx, attacker, defender, ABILITY_DAMP) == TRUE)
```
this lets us account for i.e. gastro acid much easier as well in one central place that is used wherever is necessary even in vanilla unmodified contexts
https://github.com/BluRosie/hg-engine/blob/4e312bbc85749e6e3743993c4fefe4784a651efb/src/battle/ability.c#L501-L526